Add op_vault-for-bitcoin-covenants-panel review to bitcoin-review-podcast (#637)
bitcointranscripts/bitcointranscripts: bitcoin-review-podcast/op_vault-for-bitcoin-covenants-panel.md

Reviewer: ```suggestion
Versus like with `op_vault` we can have that or like CTV or any of those, we can have it like enforced in the script that 75% will go to me and 25% to James.
```

**James O'Beirne:** 00:02:03

Hey guys.
Yeah, I'm a Bitcoin developer.
I spend a lot of my days working on core, but for the last few years, I've been really interested in custody systems and the promise of vaults and last year, I actually implemented a similar vault design on top of `OP_CTV`, which I'm sure we've covered at some point on the show, maybe.
But I was kind of dissatisfied with those implementations for various reasons, and so decided to come up with this proposal.
So just want to say that, I mean to make this a lively discussion, you should do your best to rip into this and you won't offend me, you know, kind of regardless of whatever you say, like NVK said, any new proposal for Bitcoin has to be kind of shredded apart and scrutinized.
So totally ready and welcoming of that.

**NVK:** 00:02:57

Now, this is going to be great.
So guys, there is a lot of confusion because there's a lot of stuff going on in terms of the Bitcoin Twitter drama and Bitcoin list.
I think a lot of people like me had the list muted for a few months due to, you know, tail emissions, RBF, Like, there's been a lot of stuff that happened in the last few months.

> So I got the vibe that a lot of people missed UpVault, some of the discussions that happened because it was lost in the noise and also confuses it with like, the ordinals and all this other stuff.

Reviewer: ```suggestion
So I got the vibe that a lot of people missed `OP_Vault` , some of the discussions that happened because it was lost in the noise and also confuses it with like, the ordinals and all this other stuff.
```

**Antoine Poinsot:** 00:11:59

And also the introduction of new assumptions with regards to having pre-signed transactions because if you're running let's say a network monitor, watchtower as it's called in lightning world, that would enforce your spending policies.
Let's say with `Re-Vault` basically you have a large multisig and you're delegating funds to a lower multisig.
So the covenant let's say is enforced by the fact that the lower threshold multisig does not have access to the keys of the N of N that is at the root.
So they are stuck with using an N-Vault transaction that is basically what's enforced on-chain with an open vault in more complex ways because then you have multiple keys and to this N-Vault is presigned to cancel transactions and so the N-Vault basically sends the coins either immediately to a transaction set is clawed back to the initial line of N or after a delay to some funds managers that can then use them.
In the meantime, a spending policy can be enforced by just broadcasting the canceled transaction.

> And it comes back to your initial question about how these spending policies can be enforced under Blackchain.

Reviewer: ```suggestion
And it comes back to your initial question about how these spending policies can be enforced under Blockchain.
```

But basically with pre-signed transactions, you would store them on computers and had servers and as it's stateful so you would not be able to check on your signing device and only have the signing device as a rate of trust that you can always get back your funds.
Whereas if it's encoded in a covenant, you can check it on your signing device.

**NVK:** 00:14:29

I mean, all computers are compromised.
Like, I mean, we're past that point now and the incentives are to become further compromised by even more actors compromising them, right?
Because now the money is in the computer and more people have the money in their computers.

## Anchors and fee management mechanisms

**NVK:** 00:14:44

So let's say that we have `OP_VAULT` gets in, right? and also anchors so that we can change fees.
Because you can't realistically have `OP_VAULT`s without having a way to changing fees or adding more fees.

**James O'Beirne:** 00:14:57

So what's cool now, I have yet to announce this on the mailing list, but I've actually figured out a way to avoid the reliance on v3 and anchors and all that stuff.

> **NVK:** 00:15:08

Reviewer: ```suggestion

**NVK:** 00:15:08
```

So just curious, like how do you sort of like see that?

**James O'Beirne:** 00:18:33

Yeah, so it's tricky to think about, right?
Because, you know, I think people rightfully raise the point that if in your fee management strategy you need to rely on actually having other coins available that aren't that aren't vaulted.
That's a reasonable objection to me, but at the same time, I think there are definitely plausible schemes where you've still got value locked up in that vault and so it's very easy to do this kind of atomic swap type thing where someone offers to fund your un-vault on the basis of you presenting them with an un-vault that pays them out some amount.

**NVK:** 00:19:09

I can see that as a service and I think that's not an objection.
I think this is more just trying to work out the kinks through, right?
Because it's sort of like a, we're going to have to rethink, right?
Like, and I'm speaking as somebody who does the harder wallets, right?

> Like I'm going to have to rethink of like, how do I advise the users on how they do their setups and how do they do key managemen right ?

Reviewer: ```suggestion
Like I'm going to have to rethink of like, how do I advise the users on how they do their setups and how do they do key management right ?
```

**Rijndael:** 00:19:58

The other thing that's really nice about `OP_VAULT` as opposed to other vaulting schemes that rely on pre-signed transactions is that you don't have to pre-bake them.
If you're DCAing into your savings every month or something, you can then have one un-vault or a few and then a series of recovery transactions.
That makes the fee management in the future more flexible than if every time you deposit into your savings, you have to pre-bake a transaction.
And now you have hundreds of these things and you have to worry about, how do I do fee bumping across this set of hundreds of pre-signed transactions from the 20 years that I was DCAing or something.

## Generating addresses

**NVK:** 00:20:37

How practically speaking would a user get a list of addresses or generate new addresses to deposit on a vault?

**James O'Beirne:** 00:20:44

Yeah, so you have a lot of choices there, this is kind of where we need to spend a lot of time thinking about like what kind of the recommended usage patterns are because let me describe to you like all the variability that you have in terms of what goes into an `OP_VAULT` address, right?

> So in simple terms, when you lock up coins in a vault, you're creating a `pay-to-taproot(P2TR)` transaction to a `taproot` script that looks like op-vault, and then the first parameter is the hash of your recovery path with that optional authorization key.

Reviewer: ```suggestion
So in simple terms, when you lock up coins in a vault, you're creating a `pay-to-taproot(P2TR)` transaction to a `taproot` script that looks like `OP_Vault`, and then the first parameter is the hash of your recovery path with that optional authorization key.
```

**NVK:** 00:23:33

I don't see that as an issue.
It's more a concern about, you know, people being able to link the UTXOs to the vault.

**James O'Beirne:** 00:23:40

Oh, yeah.
No. So they can't.

**NVK:** 00:23:42

Exactly.
This is beautiful that way.

## Use cases for OP_VAULT

**NVK:** 00:23:44

Yeah.
So let's say we get it in, right?
Like somehow we get it into Bitcoin, right?

**NVK:** 00:23:51

Which we should get into a little bit too, because I think it's important.
We get it in.
What do you think would be the first low hanging fruit sort of like scenarios that people are going to start building that are less complicated and like more accessible to wallet developers, right?
They may not fully comprehend it, like how far this thing can go.
So like, What are the first few case scenarios you see that the people could sort of like start using very fast?

**James O'Beirne:** 00:24:19

Yeah.

**NVK:** 00:24:20

That would be also safe.

**James O'Beirne:** 00:24:21

And I know I'll have to hand it off to Reyndall to cover because I know he's got some, but I do have a few.

> Number one is like the brain dead improvement security that everybody could make at pretty much minimal effort, which is basically just introduce, like use `Op-Vault` and have your recovery path be a sort of separate hardware manufacturer kind of wallet technique than you have right now, even if it's as stupid as like some software wallet that you spun up on a computer one time, as long as it's not correlated with the way that you store your coins right now, there's essentially no cost.

Reviewer: ```suggestion
Number one is like the brain dead improvement security that everybody could make at pretty much minimal effort, which is basically just introduce, like use `Op_Vault` and have your recovery path be a sort of separate hardware manufacturer kind of wallet technique than you have right now, even if it's as stupid as like some software wallet that you spun up on a computer one time, as long as it's not correlated with the way that you store your coins right now, there's essentially no cost.
```

I mean, the cost that you pay is like a slight delay to spend your coins, but there's no cost to introducing kind of an uncorrelated path to do recovery.
So really anybody could do that.

> The second pretty easy scenario to set up, which I think is pretty cool, is if you're worried about like a hostage situation, you can actually, and really, I mean, `Op-Vault` is kind of uniquely allows you to do this.

Reviewer: ```suggestion
The second pretty easy scenario to set up, which I think is pretty cool, is if you're worried about like a hostage situation, you can actually, and really, I mean, `Op_Vault` is kind of uniquely allows you to do this.
```

And then the family, wealth continues into this sort of rolling vault.
Is this something that you sort of like imagine or you're thinking more sort of like today, nuclear cold kind of thing?
Or it's maybe unsafe to do that, I don't know.

**James O'Beirne:** 00:27:29

I can imagine that.
I mean, you can certainly facilitate the transfer of a vault pretty securely based on the spend delay.
So if you and your family decide you want to set up different security parameters, different key configuration, obviously, like Taproot gives you a lot of flexibility in that recovery path.
So you can set up all kinds of spending conditions there.
In terms of like doing key delegation and re-negotiating, you know which keys can sign, that's, are you, Rheindell, are you unmuting this?

**Rijndael:** 00:27:56

Yeah, I was gonna say, like, the other cool thing about this being taproot is since we have Schnorr signatures, we can do things like proactive secret sharing.

> With Frost, which is a threshold signing scheme for Schnorr signatures, you can do things like you can change the quorum size of the signing set or you can change the composition of the signing set without actually doing an on-chain transaction.

Reviewer: ```suggestion
With FROST (Flexible Round-Optimized Schnorr Threshold), which is a threshold signing scheme for Schnorr signatures, you can do things like you can change the quorum size of the signing set or you can change the composition of the signing set without actually doing an on-chain transaction.
```

**Rijndael:** 00:30:50

It also expands the design space of where you can have trust-minimized third parties that can help you with management without actually delegating spend authority to them.
So James was talking about this hostage situation a minute ago.
Imagine a scenario where you're like a high net worth person or you've masked your giant stack of Bitcoin.
And so you have some company that you say like, hey, look, If you see movement out of my vault and I'm not present in your office with all of my family, then push this button and it'll trigger my emergency spend path.
And boom, that's how you deal with somebody kidnapping your kids and using it to extort your money out of you.
That's a way that you can delegate some very specific spend path execution to a third party without them actually being able to spend all of your money.
There's lots of really interesting use cases in that direction.

**NVK:** 00:31:44

Right now, there are people out there that have an envelope sealed in their lawyer's hand with their private keys.

**James O'Beirne:** 00:31:50

100%.

**NVK:** 00:31:52

And it's not a non-trivial number.

## Inheritance planning

**Rijndael:** 00:31:56

> the reason why I started talking to James about this is I think that a really underappreciated use case of vaults is for inheritance planning.

Reviewer: ```suggestion
The reason why I started talking to James about this is I think that a really underappreciated use case of vaults is for inheritance planning.
```

I think it would complete what you need to do vaults.

**James O'Beirne:** 00:59:25

So the trade off there, if you do sort of more low level granular opcodes, like TLOV or check outputs, then everybody comes up with this standardized vault construction that you have to encode in script and everybody's basically doing the same thing, but the script sizes are huge, so these transactions end up being very costly.
And I'm not even sure, to be honest, if you can get the same behavior from `OP_Vault` that you can, like, just with those two opcodes, because there is some special logic around supporting immediate re-vaults during un-vaults and things like that.
So it's, you know, I think like something like vaults is a really good example of even if you had the super futuristic flying car covenants, it's like, you still want to vault because it's a very common usage that you want to compress.

**Antoine Poinsot:** 01:00:19

Well, we don't know if it's yet a very useful, very used very common use case.
We don't know yet.

> And I don't think, well, I'm not buying that it would be so much more expensive to do with stuff to verify what you're doing with UpVault.

Reviewer: ```suggestion
And I don't think, well, I'm not buying that it would be so much more expensive to do with stuff to verify what you're doing with `OP_Vault`.
```

**James O'Beirne:** 01:00:34

It could be, but you know, nobody's shown me the scripts, like nobody's written me the scripts, like Show me the code.
People come up with these highfalutin proposals and maybe that sounds good and if I squint at it, I can interpret it in such a way, but like, where are the patches, you know?

**Rijndael:** 01:00:51

Yeah, show up with patches.

**Antoine Poinsot:** 01:00:52

Yeah.

> I agree with regard to the implementation and that's something that I was, I was wanting to, to look into, I wanted to implement a Taply Third Date Verify on ReVault, like just implementing ReVault with Taply Third Date Verify, which would basically give OP_Vault in a bit more generic way, but I never came to do it.

Reviewer: ```suggestion
I agree with regard to the implementation and that's something that I was, I was wanting to, to look into, I wanted to implement a Taproot script verify on ReVault, like just implementing ReVault with Taproot script verify, which would basically give `OP_Vault` in a bit more generic way, but I never came to do it.
```

> But AJ in this original Taply Third Date Verify post as a vault construction that is close.

Reviewer: ```suggestion
But AJ in this original Taproot script verify post as a vault construction that is close.
```

Like assuming of course there's a UI to use it and stuff, but it's like very achievable, very simple, very safe.

**James O'Beirne:** 01:02:34

Yeah, so my tagline has kind of been, you get the operational complexity of single SIG with multi-SIG safety.
And personally, if I could do this for my own funds, I would sleep a lot better at night because I could have coins in cold storage and I could be alerted if somehow in any number of a million different ways, my infrastructure got compromised, my Rodolfo backdoored my cold card.
There's just a million different ways you can get popped.
And I don't want to rely on having built the perfect fortress and pre-anticipated everything.
And so I want to be able to have reactive security.
And I think `OP_Vault` is today like kind of the most straightforward, easiest to use way to do that.

**NVK:** 01:03:24

So like, you know, essentially you get like, like end user, sort of like amazing security with like, you can build some fairly safe, straightforward scripts that handle normal people's money problems.
You can have institutional problems resolved.

**Rijndael:** 01:03:43

There's probably also a middle thing in there too.

> There's been a lot of discussion about things like e-cashments, like Cashew or Fetimint and this would also be an upgrade for their setup.

Reviewer: ```suggestion
There's been a lot of discussion about things like e-cashments, like Cashu or Fedimint and this would also be an upgrade for their setup.
```

Because what those systems are is they're basically just smaller custodians, right?

> And so if you want to uncle Jim funds for your family, your friends, your community, your school, whatever, and give them a very private lightning wallet via Chami and eCash, it would be great if you could upgrade your security setup without dramatically increasing operational complexity.

Reviewer: ```suggestion
And so if you want to uncle Jim funds for your family, your friends, your community, your school, whatever, and give them a very private lightning wallet via Chaumian eCash, it would be great if you could upgrade your security setup without dramatically increasing operational complexity.
```

**NVK:** 01:04:20

So, you get the DLCs, right?

> So we can finally have some interesting dynamic stable coins or any other kinds of dynamic contracts really being represented in the actual chain Or protected by the chain.

Reviewer: ```suggestion
So we can finally have some interesting dynamic stable coins or any other kinds of dynamic contracts really being represented in the actual chain or protected by the chain.
```

And we get some simplicity.

> I mean, like, you know, OP_Vault is a very small patch.

Reviewer: ```suggestion
I mean, like, you know, `OP_Vault` is a very small patch.
```

It's a small, I mean, compared to other things that do covenants, this is my new school.
I think, at least I have not bumped into very reasonable, very strong opposition from anybody who understands this stuff.
So I don't know, it feels like a very sane next feature for when everybody is in the mood of getting next feature.
And you resolve the fee problem, which is kind of a huge deal.
I think that the rest of the unresolved things are more practical, like how to handle the keys and how to sort of do things in the UI, sort of like the implementation real sort of like issues.
But those things are true for everything that you build new.

## Criticisms and risks

**NVK:** 01:05:31

So now like we sort of like we should start addressing like the hard part, right?
Can we go through some of like let's try to steel man some of the some of the realistic sort of criticism and risks and also maybe some of the most absurd ones.
Especially I think like addressing absurd things is very important because, you know, this stuff is not simple, right?
So like people don't get it.

**Antoine Poinsot:** 01:05:56

> Yeah, maybe if I can add something, maybe some nuances to the use case of Vaults because obviously I'm a big fan of Vult.

Reviewer: ```suggestion
Yeah, maybe if I can add something, maybe some nuances to the use case of Vaults because obviously I'm a big fan of Vault.
```

And I've been, or maybe the more nuanced that I was in the beginning after trying to solve all the issues, especially with the actual implementation and deployment of these solutions, because it's not all in the scripts and having an MVP, but trying to figure out what theories of that you're going to use and how you're going to manage this theories of and how you're going to use them is a huge problem that has huge consequences on the security of your setup as well.
So yeah, just to give more answers, is that Vaults give you spending policies.
It's give you spending policies and you can have spending policies today with a cosigner.

> So you could always have an HSM that enforces co-signing policies and you can always trust that this policy is going to be enforced as long as you trust the HSM.

Reviewer: ```suggestion
So you could always have an HSM that enforces co-signing policies and you can always trust that this policy is going to be enforced as long as you trust the HSM (Hardware Security Module).
```
Any first occurence of an abbreviation should be shown in full

**NVK:** 01:18:52

Not going to happen.

**NVK:** 01:18:55

Maybe I should just bring it up, like the CTV.
I absolutely love the work that was put into BIP119 and like, and Ruben, absolute brilliant kid.
But I think,  a lot of people don't like this, but Bitcoin is still people.
People change software.
And 100 years from now, it's going to be different people.
So they may change the software very differently than we do now.
But once you burn yourself sort of like politically to the economic nodes, which are people, I find it extremely unlikely that you can do things, especially if they're very complex and they sound like some crazy shit that people can't understand, like CTV.
So you know, drive chains changes a lot of incentives, a lot of complication on Bitcoin.
I like to call it soft fork hell.
Because essentially that's what it is.
It's just going to soft fork Bitcoin for infinitum, right?
And then CTV is like absolutely amazing, right?
But God knows, man, I mean like I don't think my brain can comprehend what can be done with that thing to be able to say with some, at least feeling confident, like it's not more than just dickbutts.

**James O'Beirne:** 01:20:03

> I think that's the thing with, with, I mean, so CTV, like you said, it's a very small patch set, but to your point, evaluating something like CTV, evaluating like `OP_CAT`, like these are very, they're simple mechanisms that can build a lot of different stuff and the conceptual surface area of what can be built is a lot higher than a proposal like check, lock, time, verify, or `OP_VAULT`, where You can fuzz test the shit out of the interface and be reasonably certain that you've like kind of exercised the full span of the space that it enables.

Reviewer: ```suggestion
I think that's the thing with, with, I mean, so CTV, like you said, it's a very small patch set, but to your point, evaluating something like CTV, evaluating like `OP_CAT`, like these are very, they're simple mechanisms that can build a lot of different stuff and the conceptual surface area of what can be built is a lot higher than a proposal like CHECKLOCKTIMEVERIFY  or `OP_VAULT`, where You can fuzz test the shit out of the interface and be reasonably certain that you've like kind of exercised the full span of the space that it enables.
```

**Ben Carman:** 01:20:57

There's just a blog post from Jeremy Rubin where he showed like you could kind of recreate drive chains with any preval.
So like, I mean, I'm a big fan of any preval.
I think there's a lot of cool things, but I am not a fan of drive chains.
So there's a lot of weird things like this.

**Antoine Poinsot:** 01:21:11

Yeah and we see it here as well.

**NVK:** 01:21:14

It's Funny, the people who have a more sort of like flexible, big composition space mindset really gets drawn to this.
Like you see this with like Fiat Jaff as well, big supporter of that, creator of Nostr, right?
And Nostr is like the complete opposite of Bitcoin.

## Unforseen consequences (ordinals)

**NVK:** 01:21:32

It's like rough consensus, but it's simple, very flexible.
That's not added a lot of rules.
Because also it's trying to do something completely different.
And I think Galaxy Brain sort of starts to get a little lost in what's acceptable to a certain purpose versus another.
And a lot of the pushback into new features and things on Bitcoin comes from this sort of like, hey, can we not break this incredible amazing thing by just adding this one more thing that we really want.

> Maybe this is a good segue to sort of start thinking like, OK, great.

Reviewer: ```suggestion
Maybe this is a good segway to sort of start thinking like, OK, great.
```

> Let's say OP_Vault, like extreme low risk, at least it is in my view, adds an incredible amount of security that Bitcoiners are going to need.

Reviewer: ```suggestion
Let's say `OP_Vault`, like extreme low risk, at least it is in my view, adds an incredible amount of security that Bitcoiners are going to need.
```

Oh, that'll be fun.

## Activating OP_VAULT

**NVK:** 01:23:02

Probably next week.
So anyways, how do we activate this?
How do we convince people?
How do we?
Because we can't even agree on how to activate shit.
And as I like to say, it should be excruciatingly horrible experience to try to activate anything, to discourage even the most hopeful people to not do it.
Because, activation is the ultimate bad attack surface of Bitcoin, right?
Like it's adding shit to it.
So like, how do we go about maybe activating this thing?

**James O'Beirne:** 01:23:30

Just a few thoughts.
I don't want to play a massive role I don't want to unilaterally push this thing obviously because that just doesn't doesn't work But you know My hope is that the value will be so obvious and and that there will be people who take time to evaluate the proposal like you, like Alex Leishman did a lot of tweeting a few days ago, the CEO of River, about how valuable this would be.
I think the people at NYDIG are pretty positive on it.
And so I just hope there's kind of like this overwhelming sense of like, wow, this is pretty low risk and pretty high value.
And it's something that we want.

> And the other note that I'd like to make is like there was a time where for things like check clock time verify, check sequence verify, like these were like purpose specific tools and their activation wasn't full of drama, wasn't this huge massive thing that people freaked out about.

Reviewer: ```suggestion
And the other note that I'd like to make is like there was a time where for things like CHECKLOCKTIMEVERIFY, CHECKSEQUENCEVERIFY like these were like purpose specific tools and their activation wasn't full of drama, wasn't this huge massive thing that people freaked out about.
```

And that might be a more compelling argument to the economic majority than just saying like here's a really great idea that I have.

**Antoine Poinsot:** 01:27:14

Yeah I agree with that it would be useful for arrivals, but two things, first, I really don't want to be into pushing for consensus changes for my company.
And two, I actually don't think we should, well, not rush, but I don't think we should go with activating anything not too soon.
So there's been a lot of covenant proposals floating around.
People are still working on covenant proposals that are not announced yet.
There is a lot of research going into it.
So I think it would be premature to activate anything.

**NVK:** 01:27:48

What would be the least disruptive way of getting this in?
Even if it was like maybe like not fully featured?

**Ben Carman:** 01:27:56

I think something to keep in mind as well is like we don't have to do `OP_Vault` and that's it for Covenants.
Like we could do `OP_Vault` and `OP_CTV` and APO.

**NVK:** 01:28:03

No, I know

**Ben Carman:** 01:28:04

It's not going to be all at once, Yeah.

**NVK:** 01:28:06

> But like The way that the people like if you use some EQ here, right, A lot's true.

Reviewer: A different speaker

Insurance is basically not doable.
The situation right now is pretty bad, and we're all used to it, and so maybe we don't realize how bad it is, but it could be a lot better.
So I'm not really wild about the prospect of waiting three years for some ultra-perfect covenant proposal to come along that like maybe kind of emulates `OP_Vault`.
You know, I think I'd like to put the proposal out there, see what the objections are.
And if someone is like, hey, I really don't like that we can do this DLC thing with the pack.

**NVK:** 01:30:46

You know, if I have to like, sort of like really address what I think it's gonna happen, is that it's not gonna necessarily, I don't think the criticism is gonna come regarding the actual nuances of `OP_Vault`.
It's gonna be activation.
Is the idea is like, the problem is, I think Speedy Trial like pissed off enough people.
And it was kind of like felt rammed through by the cabal of core developers.
So there is that bad taste in the mouth of a lot of economic nodes out there.
The dick butts certainly did not help.
And I still think we don't have a good way of activating stuff.
And we sort of like, we're a little PTSD from things.

> So it's not going to be OP_Vault.

Reviewer: ```suggestion
So it's not going to be `OP_Vault`.
```

And you need the people who understand the code base to sort of upgrade it for new hardware, for new concerns, for, Unix time problems.
And there's all these things that need to change.
And you know, when my good friend Steve says, fire the devs, I understand why it feels that way.
And I kind of agree in a way because, like core, I addressed this in the last episode.
It kind of feels like this, it's like Internet Explorer has 99% of the market.
It doesn't rule the internet, but kind of rules the internet.
And, it has this sort of like bad taste by the CTV people as well, where, hey, if you don't get the right people on court to like your idea, It doesn't get activated to you.
So these are the gatekeepers.
But you know, like realistically speaking in software, like there's always just like a literal handful of people who understand the whole banana and are the people who are qualified to truly have an opinion.
You know, like Facebook used to have five guys merging the code of 5,000 people.
You know, in our shop, like we have, you know, Doc Hex and if he doesn't get through his shit, like it's not going to get through.

> And very few people have the full picture of that cold base.

Reviewer: ```suggestion
And very few people have the full picture of that codebase.
```

And those people are often not the best people who are selling themselves or selling the cold.
So I think like Bitcoin is sort of like suffering at this right now.

> Like that's the current sort of like cold and sneeze that Bitcoin has.

Reviewer: ```suggestion
Like that's the current sort of like code and sneeze that Bitcoin has.
```

Years ago.

**Antoine Poinsot:** 01:39:40

Yeah, a decade of research, a lot of involvement of a lot of people in the community with the workshops, a lot of involvement with the activation, and maybe that's why as well people are happy with it because they were actually involved and so, yeah, I think that did put not a very high bar, but reasonable bar for community involvement and normal user involvement.

**Rijndael:** 01:40:03

Well, like schnorr is injured for a similar thing.
Like people have been wanting schnorr in Bitcoin for years.

**NVK:** 01:40:08

See, like, you know, the schnorr is funny that you brought that up.
Like, you know, it's always been a concern of mine that like, what if `ECDSA` is backdoor right?
And it's not revealed for a long time and there is no proof for for `ECDSA`.
So, you know, the idea of like, and again, this is a fundamental change to Bitcoin, right.
Adding another crypto primitive, it's like crazy different.
So, You know, but just having a secondary fallback crypto primitive in Bitcoin before we're big enough to state actors are going to that extent.
It's pretty cool.
Thanks Antoine, Antoine is stepping out.
Appreciate it.

> But now, like, you know, we have this sort of like the fights have always been hard, you know, up return limit back in 2009, 2010, because people were concerned about the worst part of dick butts.

Reviewer: ```suggestion
But now, like, you know, we have this sort of like the fights have always been hard, you know, `OP_Return` limit back in 2009, 2010, because people were concerned about the worst part of dick butts.
```
